### PR TITLE
CMP: add support for offline client

### PR DESCRIPTION
**Pull request by DDvO**
This is helpful in situations where a CMP client does not (yet) have an online connection to PKI services for requesting a certificate.
The client dumps its request message using `-reqout` to a file without attempting to contact a server.
Later the client, or even some proxy not having access to the private key to be certified
(which usually is needed for the proof of possession), can pick up request message using `-reqin` and send it out.
Upon receiving the response, it can verify the response and check the included certificate
by matching the certified public key with the public key in the request message file.
If this done by a proxy, the response message can also be dumped using `-rspout` and transferred back to the client,
which the client can read in using `-rspin`, verify it, and check the included certificate - 
if done in conjunction with `-reqin`, the client does not need to access its own newly certified private key in this step.

To this end,
* `apps/cmp`: add `-reqout_only` option for dumping/saving just the initial CMP request message, 
which is like `-reqout` but does not contact any server
* `apps/cmp`: extend documentation and diagnostics for using `-reqin` in special situations
*  `apps/cmp`: improve `-reqin` option to read fallback public key from first request message file given; to this end:
* `crypto/cmp/cmp_msg.c`: add `OSSL_CMP_MSG_get0_certreq_publickey()` and fix a coding style nit

On this occasion, refactor `apps/cmp.c` a little to fix some coding style nits and for more convenient source-level debugging (by factoring out the first loop over the CLI arguments to a helper function `handle_opts_upfront()`).

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
Rebased to fix trivial merge conflict in `apps/cmp.c`.

**Comment by DDvO** [comment]
rebased to fix merge conflicts

**Comment by DDvO** [comment]
Rebased again to fix merge conflicts.

Can we please get this done now as well?

**Comment by DDvO** [comment]
Rebased to fix further merge conflict

File: crypto/cmp/cmp_msg.c [hunk 1]
```diff
@@ -100,6 +100,34 @@ int OSSL_CMP_MSG_get_bodytype(const OSSL_CMP_MSG *msg)
...
+    case OSSL_CMP_PKIBODY_KUR:
+        reqs = msg->body->value.ir; /* value.ir is same for cr and kur */
+        if ((crm = sk_OSSL_CRMF_MSG_value(reqs, 0)) == NULL) {
+            ERR_raise(ERR_LIB_CMP, CMP_R_CERTREQMSG_NOT_FOUND);
+            return NULL;
+        }
+        if ((tmpl = OSSL_CRMF_MSG_get0_tmpl(crm)) == NULL
+            || (pubkey = OSSL_CRMF_CERTTEMPLATE_get0_publicKey(tmpl)) == NULL) {
+            ERR_raise(ERR_LIB_CMP, CRMF_R_POPO_MISSING_PUBLIC_KEY);
+            return 0;
```
Comment (t8m) [inline crypto/cmp/cmp_msg.c:122]: `return NULL;`

**Review by t8m — CHANGES_REQUESTED** [review]
Just 3 minor things

File: apps/cmp.c [hunk 2]
```diff
@@ -814,6 +819,14 @@ static OSSL_CMP_MSG *read_write_req_resp(OSSL_CMP_CTX *ctx,
     OSSL_CMP_PKIHEADER *hdr;
     const char *prev_opt_rspin = opt_rspin;
 
+    if (opt_reqout_only != NULL) {
+        if (OSSL_CMP_MSG_write(opt_reqout_only, req) < 0)
+            CMP_err1("cannot write request PKIMessage to file '%s'",
+                     opt_reqout_only);
+        else
+            reqout_only_done = 1;
+        return 0; /* in any case, stop at this point, not contacting server */
```
[the remark this answers is not in the document]
Comment (DDvO) [reply apps/cmp.c]: Oops, fixed.

File: crypto/cmp/cmp_msg.c (hunk 1 above)
Comment (DDvO) [reply crypto/cmp/cmp_msg.c]: Fixed as well.

File: crypto/err/openssl.txt [hunk 3]
```diff
@@ -278,7 +278,7 @@ CMP_R_UNEXPECTED_PKIBODY:133:unexpected pkibody
 CMP_R_UNEXPECTED_PKISTATUS:185:unexpected pkistatus
 CMP_R_UNEXPECTED_POLLREQ:105:unexpected pollreq
 CMP_R_UNEXPECTED_PVNO:153:unexpected pvno
-CMP_R_UNEXPECTED_SENDER:197:unexpected sender
+CMP_R_UNEXPECTED_SENDER:106:unexpected sender
```
[the remark this answers is not in the document]
Comment (t8m) [reply crypto/err/openssl.txt:281]: Hmm, OK. Although numbers are cheap :grin:

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Thank you Tomas for reviewing and approving this.

Asking @openssl/committers for 2nd review.

**Comment by DDvO** [comment]
Ping @openssl/committers for 2nd review.

**Comment by DDvO** [comment]
@openssl/committers could anyone please give his blessing,
such that this makes it into 3.3, after already waiting in here since August.

**Review by beldmit — APPROVED** [review]
I don't understand the purpose of the code but have overall impression it is in a good shape

**Comment by DDvO** [comment]
Merged - thanks a lot @t8m and @beldmit for your reviews

**Pull request closed without merging**